src/kmeans.jl

```diff
@@ -269,89 +280,82 @@ function update_centers!(
     end
 
     # sum ==> mean
-    @inbounds for j = 1:k
+    @inbounds for j ∈ 1:k
         if to_update[j]
-            for i = 1:d
-                centers[i, j] /= cweights[j]
+            cj = cweights[j]
+            for i ∈ 1:d
+                centers[i, j] /= cj
             end
         end
     end
 end
 
-#
-#  Update centers based on updated assignments
-#
-#  (specific to the case where samples are weighted)
-#
-function update_centers!(
-    X::AbstractMatrix{T},            # in: sample matrix (d x n)
-    weights::AbstractVector{<:Real}, # in: sample weights (n)
-    assignments::Vector{Int},        # in: assignments (n)
-    to_update::Vector{Bool},         # in: whether a center needs update (k)
-    centers::AbstractMatrix{T},      # out: updated centers (d x k)
-    cweights::Vector{Float64}        # out: updated cluster weights (k)
-    ) where T<:Real
+####
+#### update centers (weighted case)
+####
 
+function update_centers!(X::AbstractMatrix{TX}, weights::Vector{TW},
+                         assignments::Vector{Int}, to_update::Vector{Bool},
+                         centers::Matrix{TC}, cweights::Vector{TW}
+                         ) where {TX<:Real,TC<:AbstractFloat,TW<:Real}
     d, n = size(X)
     k = size(centers, 2)
 
     # initialize center weights
-    cweights[to_update] .= 0.0
+    cweights[to_update] .= zero(TW)
```

Comment:
This should really make no difference in practice. Anyway better use `0` with is the simplest zero type to reason about for the compiler.

Reply:
ok